# Preprocessing

This is the first R markdown script focused in data preparation

## Import the dataset

Import the datset in our R workspace.

```{r, eval = F}
# Set the path of the data directory
path <- "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data"

# Set the path for Analysis output
output <- "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis"

# Set the path of the Raw data
Raw_data <- file.path(path, "Raw_Data/RUN1")

# Check if all the files are present
list.files(Raw_data)
```

We will further need the name of our sample.
```{r, eval = F}
# list all forward and reverse reads in these objects
fnFs <- sort(list.files(Raw_data, pattern = "R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(Raw_data, pattern = "R2_001.fastq.gz", full.names = TRUE))

# Create also an object containing the ID numbers
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 2)
```

## Data preparation

The primers need to be removed. The primers used by SciLifeLab are 341F and 805R primers.

```{r, eval = F}
# Set up primers object
FWD <- "CCTACGGGNGGCWGCAG"
REV <- "GACTACHVGGGTATCTAATCC"
```

To remove the primers, it is advised by SciLifeLab to use Cutadapt.

Cutadapt has been set in my /bin dorectory in my case so no need to set a path

```{r, eval = F}
# Set output directory for the data
cut_output <- file.path(path, "Trimmed/")
if (!dir.exists(cut_output)) dir.create(cut_output)

# Set path for the output file
fnFs_cut <- file.path(cut_output, basename(fnFs))
fnRs_cut <- file.path(cut_output, basename(fnRs))

# Check the cutadapt version
system2("/opt/anaconda3/bin/cutadapt", args = "--version")

# We need to create a loop to remove in every files the primer
for (i in seq_along(fnFs)) {
  system2("/opt/anaconda3/bin/cutadapt",
    args = c(
      "-g", FWD, "-G", REV,
      "--discard-untrimmed",
      "-o", fnFs_cut[i], "-p", fnRs_cut[i],
      fnFs[i], fnRs[i]
    )
  )
}

# Check if the trimming worked
head(list.files(cut_output))
```

After Trimming the primers we can do a Quality check using FastQC and MultiQC.

## Quality check

```{r, eval = F}
# Create a directory in the analysis for the quality control
if (!dir.exists(file.path(output, "Quality_Control"))) dir.create(file.path(output, "Quality_Control"))

# Run fastqc in all the samples
system2("fastqc", args = c("-o", file.path(output, "Quality_Control"), file.path(cut_output, "*")))

# Merge them with multiqc
system2("multiqc", args = c(file.path(output, "Quality_Control")))
```

Now That the data got prepared by removing the primers, we can go to further analysis.

The entire workspace can be saved for later.
```{r, eval = F}
# Set the path for our backup workspaces
save <- "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Backup_workspace"

# Save the workspace
save.image(file.path(save, "Data_preparation.RData"))
```