# Decomtamination and normalization step pipeline

## library

```{r, eval = F}
# Analysis package
library(phyloseq)
library(Biostrings)
packageVersion("Biostrings")
library(decontam)
library(ggplot2)
library(vegan)
library(tidyverse)
library(decontam)
packageVersion("decontam")
```

## Load Data

```{r, eval = F}
load("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Backup_workspace/DaDa2_pipeline_RUN1.RData")

# Load the cleaned metadata as well
load("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Backup_workspace/Cleaned_Metadata.RData")
```

## Construct the phyloseq object

```{r, eval = F}
# Because some duplicates in the Samples remove them from Seqtab + Non present in Genomic data
seqtab.nochim <- seqtab.nochim[!(row.names(seqtab.nochim) %in% c("157", "365", "235")), ]

# Order Metadata file based on Sample Names
Metadata_cleaned_microbiome <- Metadata_cleaned_microbiome[order(Metadata_cleaned_microbiome$Sample.Name), ]

# Change row Names
rownames(Metadata_cleaned_microbiome) <- rownames(seqtab.nochim)

# Remove Sample Names from Metadata file
Metadata_cleaned_microbiome <- Metadata_cleaned_microbiome[-c(2)]
# Create the ps object
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = F), sample_data(Metadata_cleaned_microbiome), tax_table(taxa))

# Reformat it
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

## Decontamination steps

```{r, eval = F}
# Contaminants Prevalence filtering
sample_data(ps)$is.neg <- sample_data(ps)$Sample.Type == "Blanks"
contamdf.prev <- isContaminant(ps, method = "prevalence", neg = "is.neg", threshold = 0.5)
table(contamdf.prev$contaminant)
head(which(contamdf.prev$contaminant))

# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(ps, function(abund) 1 * (abund > 0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$Sample.Type == "Blanks", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Sample.Type != "Blanks", ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(
    pa.pos = taxa_sums(ps.pa.pos), pa.neg = taxa_sums(ps.pa.neg),
    contaminant = contamdf.prev$contaminant
)

ps_noncontam <- prune_taxa(!contamdf.prev$contaminant, ps)

# Remove is.neg column
ps_noncontam <- subset_samples(ps_noncontam, select = -is.neg)
sample_data(ps_noncontam)

ps <- subset_samples(ps_noncontam, Sample.Type != "Blanks")
```

```{r, eval = F}
head(tax_table(ps)) # check
get_taxa_unique(ps, "Family")
get_taxa_unique(ps, "Class")
get_taxa_unique(ps, "Order")

mitochondria <- phyloseq::subset_taxa(ps, Family == "Mitochondria")
mitochondria
sample_sums(mitochondria)[order(sample_sums(mitochondria))]

chloroplast <- subset_taxa(ps, Order == "Chloroplast")
chloroplast
tax_table(chloroplast)[, 1:7]
sample_sums(chloroplast)[order(sample_sums(chloroplast))]

# Eukaryotes and unassigned (I don't have Eukaryotes in my samples)
unassigned <- subset_taxa(ps, is.na(Kingdom))
unassigned
tax_table(unassigned)[, 1:7]
sample_sums(unassigned)[order(sample_sums(unassigned))]

# filter mitochondria
badTaxa <- row.names(tax_table(mitochondria)[, 1])
allTaxa <- taxa_names(ps)
goodTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
ps <- prune_taxa(goodTaxa, ps)
# filter chloroplast
badTaxa <- row.names(tax_table(chloroplast)[, 1])
allTaxa <- taxa_names(ps)
goodTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
ps <- prune_taxa(goodTaxa, ps)
# filter Eukaryote reads
badTaxa <- row.names(tax_table(unassigned)[, 1])
allTaxa <- taxa_names(ps)
goodTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
ps <- prune_taxa(goodTaxa, ps)

# filter zero ASVs
table(taxa_sums(ps) < 1)
ps
ps <- prune_taxa(taxa_sums(ps) > 0, ps)
```


## Rarefaction filtering (Will check if possible)

```{r, eval = F}
# Plot the rarefaction curve
ps_rarecurve <- prune_samples(sample_sums(ps) <= 10000, ps)

tab <- otu_table(ps_rarecurve)
class(tab) <- "matrix"
rare <- rarecurve(tab, step = 100000, lwd = 2, ylab = "OTU", label = F)

# Normalize and filter low count reads

ps_Norm <- transform_sample_counts(ps, function(x) x / sum(x))
ps_Norm_filt <- filter_taxa(ps_Norm, function(x) mean(x) > 1e-5, TRUE)

# Rarefy reads
set.seed(200)
ps_filt <- prune_taxa(taxa_names(ps_Norm_filt), ps)
ps_rare <- rarefy_even_depth(ps, sample.size = min(sample_sums(ps)))
ps_rare_filt <- rarefy_even_depth(ps_filt, sample.size = min(sample_sums(ps_filt)))
```

## Save the workspace

```{r, eval = F}
# Erase everything not needed anymore
rm(list = ls()[!(ls() %in% c("seqtab.nochim", "taxa", "save", "ps", "ps_rare", "ps_rare_filt", "ps_Norm", "rare", "ps_filt", "ps_Norm_filt", "Metadata_cleaned_microbiome"))])

# Save workspace
save.image(file.path(save, "Decomtamination_pipeline_RUN1.RData"))
```