# Relative Abundance Analysis

This script contains relative Abundance visualizations in different aspect

## Library

```{r, eval = F}
library("phyloseq")
packageVersion("phyloseq")
library("ggplot2")
packageVersion("ggplot2")
library(dplyr)
packageVersion("dplyr")
```
## Load the dataset

```{r, eval= F}
load("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Backup_workspace/Decomtamination_pipeline_RUN1.RData")
```

## Set Theme

```{r, eval = F}
theme_set(theme_bw())
```

## Summary statistics of percentage (Phylum/ Genus)

```{r}
# create melted object
melted_ps_object <- psmelt(ps)

# check the melted object phylum
table(melted_ps_object$Phylum, useNA = "ifany")

# Calculate the sum of reads per phylum
Phylum_read_count <- aggregate(melted_ps_object$Abundance, by = list(Category = melted_ps_object$Phylum), FUN = sum)
Phylum_read_count <- Phylum_read_count[order(-Phylum_read_count$x), ]
Phylum_read_count

# Change in percentage
Phylum_read_count$x <- (Phylum_read_count$x/sum(Phylum_read_count$x)*100)
Phylum_read_count

# Calculate the mean of reads per Genus
Genus_read_count <- aggregate(melted_ps_object$Abundance, by = list(Category = melted_ps_object$Genus), FUN = sum)
Genus_read_count <- Genus_read_count[order(-Genus_read_count$x), ]
Genus_read_count

# Change in percentage
Genus_read_count$x <- (Genus_read_count$x/sum(Genus_read_count$x)*100)
Genus_read_count
```

## Phylum Relative abundance stacked barplot

```{r, eval = F}
# Reduce to the Phylum level
Agglom_Phylum <- tax_glom(ps, taxrank = "Phylum")
Agglom_Phylum <- transform_sample_counts(Agglom_Phylum, function(x) x/sum(x))
barplot_phylum <- psmelt(Agglom_Phylum)
table(barplot_phylum$Phylum, useNA = "ifany") # table of phyla
barplot_phylum$Phylum[barplot_phylum$Phylum == ""] <- "NA" # change if there are empty names to NA
table(barplot_phylum$Phylum, useNA = "ifany") # check

# calculate sum of abundances
taxa_summary <- aggregate(barplot_phylum$Abundance, by = list(Category = barplot_phylum$Phylum), FUN = mean)
taxa_summary <- taxa_summary[order(-taxa_summary$x), ]
taxa_summary
```

### Select top 5 Phyla and make the rest labeled as "other"

```{r, eval = F}
list(as.character(taxa_summary[c(1:5), 1]))

top_phyla <- as.character(c("Bacillota", "Bacteroidota", "Pseudomonadota", "Actinomycetota", "Verrucomicrobiota"))

# Function
"%!in%" <- function(x, y) !("%in%"(x, y))

# For-loop and if then function that checks to see if any of the names in the list "top" is in the
# barplot_phylum$Phylum column, leaves unchanged if it is, changes it to "Other" if it isn't
# Re-name all uncommon phyla as "Other"
for (i in seq_along(barplot_phylum$Phylum)) {
      if (barplot_phylum$Phylum[i] %!in% top_phyla) {
            barplot_phylum$Phylum[i] <- "Other"
      }
}

# Re-name NA:s as "Other"
barplot_phylum$Phylum[is.na(barplot_phylum$Phylum)] <- "Other"
```

Change order so "Other" is last, change the rest to whichever order you prefer:

```{r, eval = F}
barplot_phylum$Phylum <- factor(barplot_phylum$Phylum, levels = c("Bacillota", "Bacteroidota", "Pseudomonadota", "Actinomycetota", "Verrucomicrobiota", "Other"))
levels(barplot_phylum$Phylum)
```

### Relative Abundance by samples

```{r, eval = F}
# Reorder based on Bacilllota
barplot_phylum %>% 
  group_by(ID) %>% 
  summarise(Sum = sum(Abundance)) %>% 
  arrange(Sum) %>% 
  print(n = Inf)

barplot_phylum$ID_new <- as.character(barplot_phylum$ID)

ID_new_levels <-
  barplot_phylum %>%
  filter(Phylum == "Bacillota")  %>%
  group_by(ID_new)%>% 
  summarise(Sum = sum(Abundance)) %>%
  arrange(Sum) %>%
  pull(ID_new) %>%
  unique

barplot_phylum$ID_new <- factor(barplot_phylum$ID_new, levels = ID_new_levels)

ggplot(barplot_phylum, aes(x = ID_new, y = Abundance, order = Phylum)) +
    geom_bar((aes(fill = Phylum)), stat = "identity", position = "stack", width = 1) +
    facet_wrap(~Island, scales = "free_x") +
    ylab("Relative abundance") +
    theme_bw() +
    theme(panel.grid.minor=element_blank(), 
          panel.grid.major=element_blank(),
          panel.background = element_blank()) +
    theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=13),
        axis.text.y = element_text(size=12, color ="black")) +
    theme(legend.title = element_text(size=14),
          legend.text = element_text(size=12),
          legend.key.size=unit(0.7,"cm"),
          legend.position = "bottom",
          legend.box.margin = margin(5, 5, 5, 5),
          legend.box.background = element_rect()) 
# scale_fill_manual(values=(color_list[c(10,6,8,5,4,3)])) +
# facet_grid(~Sample_type, scales="free", space="free") + # change this per your liking!
# scale_y_continuous(expand = c(0,0)) +
# theme(strip.text.x = element_text(size = 8, color = "black"),
# strip.background = element_rect(fill = "#ffffff"))
```

## Relative Abundance at genus level (Overall /core Microbiome)

### Top Genus check

```{r, eval = F}
# Reduce to the Phylum level
Agglom_Genus <- tax_glom(ps, taxrank = "Genus")
Agglom_Genus <- transform_sample_counts(Agglom_Genus, function(x) x/sum(x))
barplot_Genus <- psmelt(Agglom_Genus)
table(barplot_Genus$Genus, useNA = "ifany") # table of phyla
barplot_Genus$Genus[barplot_Genus$Genus == ""] <- "NA" # change if there are empty names to N

table(barplot_Genus$Genus, useNA = "ifany") # check

# calculate mean of relative abundances
taxa_summary <- aggregate(barplot_Genus$Abundance, by = list(Category = barplot_Genus$Genus), FUN = mean)
taxa_summary$Mean_Relative_Abundance <- taxa_summary$x
taxa_summary <- taxa_summary[order(-taxa_summary$Mean_Relative_Abundance), ]
taxa_summary
```

### Select the Genus under 1% mean relative abundance and change to others

```{r, eval = F}
# Calculate RElative Abundance mean
Mean_Rel_Genus <- aggregate(barplot_Genus$Abundance, by = list(Category = barplot_Genus$Genus), FUN = mean)

# list of the Remainder Genus
Remain_Genus <- Mean_Rel_Genus[Mean_Rel_Genus$x <= 0.01,]$Category

# Change their Name to "other"
barplot_Genus[barplot_Genus$Genus %in% Remain_Genus,]$Genus <- 'Other genus <1%'

print(levels(factor(barplot_Genus$Genus)))
```

```{r, eval = F}
list_Genus  <- unique(barplot_Genus$Genus)
list_Genus <- list_Genus[list_Genus != "Other genus <1%"]
list_Genus <- list_Genus[list_Genus != "Bacteroides"]
barplot_Genus$Genus <- factor(barplot_Genus$Genus, levels = c("Bacteroides", list_Genus, "Other genus <1%") )
                             
levels(barplot_Genus$Genus)
```

### Relative Abundance by samples

```{r, eval = F}
# Reorder based on Bacilllota
barplot_Genus %>% 
  group_by(ID) %>% 
  summarise(Sum = sum(Abundance)) %>% 
  arrange(Sum) %>% 
  print(n = Inf)

barplot_Genus$ID_new <- as.character(barplot_Genus$ID)

ID_new_levels <-
  barplot_Genus %>%
  filter(Genus == "Bacteroides")  %>%
  group_by(ID_new)%>% 
  summarise(Sum = sum(Abundance)) %>%
  arrange(Sum) %>%
  pull(ID_new) %>%
  unique

barplot_Genus$ID_new <- factor(barplot_Genus$ID_new, levels = ID_new_levels)

ggplot(barplot_Genus, aes(x = ID_new, y = Abundance, order = Genus)) +
    geom_bar((aes(fill = Genus)), stat = "identity", position = "stack", width = 1) +
    facet_grid(~Island, scales = "free_x", space = "free") +
    ylab("Relative abundance") +
    theme_bw() +
    theme(panel.grid.minor=element_blank(), 
          panel.grid.major=element_blank(),
          panel.background = element_blank()) +
    theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=13),
        axis.text.y = element_text(size=12, color ="black")) +
    theme(legend.title = element_text(size=14),
          legend.text = element_text(size=12),
          legend.key.size=unit(0.7,"cm"),
          legend.position = "none",
          legend.box.margin = margin(5, 5, 5, 5),
          legend.box.background = element_rect())
    #scale_fill_manual(values = others_col, limits = c("Other genus <1%"))
# scale_fill_manual(values=(color_list[c(10,6,8,5,4,3)])) +
# facet_grid(~Sample_type, scales="free", space="free") + # change this per your liking!
# scale_y_continuous(expand = c(0,0)) +
# theme(strip.text.x = element_text(size = 8, color = "black"),
# strip.background = element_rect(fill = "#ffffff"))
```

### Core Microbiome Relatives Abundance Plot

```{r, eval = F}
# Reduce to the Phylum level
Agglom_Genus_core <- core(Agglom_Genus, detection = 0, prevalence = 0.85)
barplot_Genus_core <- psmelt(Agglom_Genus_core)
table(barplot_Genus_core$Genus, useNA = "ifany") # table of phyla
barplot_Genus_core$Genus[barplot_Genus_core$Genus == ""] <- "NA" # change if there are empty names to N

table(barplot_Genus_core$Genus, useNA = "ifany") # check

# calculate mean of relative abundances
taxa_summary <- aggregate(barplot_Genus_core$Abundance, by = list(Category = barplot_Genus_core$Genus), FUN = mean)
taxa_summary$Mean_Relative_Abundance <- taxa_summary$x
taxa_summary <- taxa_summary[order(-taxa_summary$Mean_Relative_Abundance), ]
taxa_summary
```

### Select the Genus under 1% mean relative abundance and change to others

```{r, eval = F}
# Calculate RElative Abundance mean
Mean_Rel_Genus_core <- aggregate(barplot_Genus_core$Abundance, by = list(Category = barplot_Genus_core$Genus), FUN = mean)

# list of the Remainder Genus
Remain_Genus_core <- Mean_Rel_Genus_core[Mean_Rel_Genus_core$x <= 0.01,]$Category

# Change their Name to "other"
barplot_Genus_core[barplot_Genus_core$Genus %in% Remain_Genus_core,]$Genus <- 'Other genus <1%'

print(levels(factor(barplot_Genus_core$Genus)))
```

```{r, eval = F}
list_Genus  <- unique(barplot_Genus_core$Genus)
list_Genus <- list_Genus[list_Genus != "Other genus <1%"]
list_Genus <- list_Genus[list_Genus != "Bacteroides"]
barplot_Genus_core$Genus <- factor(barplot_Genus_core$Genus, levels = c("Bacteroides", list_Genus, "Other genus <1%") )
                             
levels(barplot_Genus_core$Genus)
```

### Relative Abundance by samples

```{r, eval = F}
# Reorder based on Bacilllota
barplot_Genus_core %>% 
  group_by(ID) %>% 
  summarise(Sum = sum(Abundance)) %>% 
  arrange(Sum) %>% 
  print(n = Inf)

barplot_Genus_core$ID_new <- as.character(barplot_Genus_core$ID)

ID_new_levels <-
  barplot_Genus_core %>%
  filter(Genus == "Bacteroides")  %>%
  group_by(ID_new)%>% 
  summarise(Sum = sum(Abundance)) %>%
  arrange(Sum) %>%
  pull(ID_new) %>%
  unique

barplot_Genus_core$ID_new <- factor(barplot_Genus_core$ID_new, levels = ID_new_levels)


ggplot(barplot_Genus_core, aes(x = ID_new, y = Abundance, order = Genus)) +
    geom_bar((aes(fill = Genus)), stat = "identity", position = "stack", width = 1) +
    facet_grid(~Island, scales = "free_x", space = "free") +
    ylab("Relative abundance") +
    theme_bw() +
    theme(panel.grid.minor=element_blank(), 
          panel.grid.major=element_blank(),
          panel.background = element_blank()) +
    theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=13),
        axis.text.y = element_text(size=12, color ="black")) +
    theme(legend.title = element_text(size=14),
          legend.text = element_text(size=12),
          legend.key.size=unit(0.7,"cm"),
          legend.position = "none",
          legend.box.margin = margin(5, 5, 5, 5),
          legend.box.background = element_rect())
# scale_fill_manual(values=(color_list[c(10,6,8,5,4,3)])) +
# facet_grid(~Sample_type, scales="free", space="free") + # change this per your liking!
# scale_y_continuous(expand = c(0,0)) +
# theme(strip.text.x = element_text(size = 8, color = "black"),
# strip.background = element_rect(fill = "#ffffff"))
```

### Identify the outliers

```{r}
# Extract the Samples with extremely low sum of relative abundance
Sum_Rel_Genus_core_sample <- aggregate(barplot_Genus_core$Abundance, by = list(Category = barplot_Genus_core$ID), FUN = sum)

Sum_Rel_Genus_core_sample
Sum_Rel_Genus_core_sample <- Sum_Rel_Genus_core_sample[order(Sum_Rel_Genus_core_sample$x), ]
hist(Sum_Rel_Genus_core_sample$x, breaks = 100, main = "Histogram Sum of Relative Abundance per individuals", xlab = "Sum of Relative Abundance")
Sum_Rel_Genus_core_sample
```

