# Script for Population Composition visualization
This script is gonna be used mostly to get visual representation of Sample composition. ex. Bermuda/Hawai , Location, male/Femal, population stratification, etc.

## library
```{r}
library(readr)
library(ggplot2)
library(ppcor)
library(corrplot)
library(dplyr)
```

## Load the metadata/covariate file
```{r}
# Load the microbiome metadata file
metadata <- read.csv("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/Metadata/Metadata_Chicken.csv", na = "NA")

# Remove all the NA values
metadata <- metadata[!is.na(metadata$Sample.Name), ]

# Sort by sample Name
metadata <- metadata[order(metadata$Sample.Name), ]

# Match metadata file with seqtab (for the first run only)
metadata <- metadata[!(row.names(metadata) %in% c("450", "451", "452")), ]

## Load the second file
feral_covariates_microbiome <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/feral_covariates_microbiome.txt", na = "NA")

# Reformat
feral_covariates_microbiome$Island <- replace(feral_covariates_microbiome$Island, feral_covariates_microbiome$Island == "kauai", "Kauai")
feral_covariates_microbiome$Island <- replace(feral_covariates_microbiome$Island, is.na(feral_covariates_microbiome$Island), "Kauai")

# Change the Sex column
feral_covariates_microbiome$calc_sex <- with(feral_covariates_microbiome, ifelse(grepl("f", feral_covariates_microbiome$calc_sex, ignore.case = T, fixed = T), "Female", "Male")) # nolint: T_and_F_symbol_linter.

# Merge both metadata set togheter
Metadata_cleaned_microbiome <- merge(metadata[c(1, 3:4)], feral_covariates_microbiome, by = "ID")

# Import the Genomic PCA 
Genome_eigenvec <- read.table("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/PLINK/PCA/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink-job9079824.PCA-job9667882.eigenvec", header = FALSE)

# Remove the second column
Genome_eigenvec <- Genome_eigenvec[, !(names(Genome_eigenvec) %in% c("V2"))]

# Rename first column
colnames(Genome_eigenvec) <- c("ID","PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10","PC11","PC12","PC13","PC14","PC15","PC16","PC17","PC18","PC19","PC20")

# Add the Genomic PC information to Metadata file
Metadata_cleaned_microbiome <- merge(Metadata_cleaned_microbiome[, !(names(Metadata_cleaned_microbiome) %in% c("PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10","PC11","PC12","PC13","PC14","PC15","PC16","PC17","PC18","PC19","PC20"))], Genome_eigenvec, by = "ID")

## Remove Unecessery information

# remove Latitude, Longitude, latitude.longitude
Metadata_cleaned_microbiome <- Metadata_cleaned_microbiome[, !(names(Metadata_cleaned_microbiome) %in% c("Latitude", "Longitude", "latitude.longitude"))]

## Format in the good type
Metadata_cleaned_microbiome$location.numeric <- as.factor(Metadata_cleaned_microbiome$location.numeric)

Metadata_cleaned_microbiome$sex_calc_numeric <- as.factor(Metadata_cleaned_microbiome$sex_calc_numeric)

Metadata_cleaned_microbiome$Age_category <- as.factor(Metadata_cleaned_microbiome$Age_category)

Metadata_cleaned_microbiome$Island <- as.factor(Metadata_cleaned_microbiome$Island)

Metadata_cleaned_microbiome$Sample.Type <- as.factor(Metadata_cleaned_microbiome$Sample.Type)

Metadata_cleaned_microbiome$calc_sex <- as.factor(Metadata_cleaned_microbiome$calc_sex)
```

###  Create a dataset based only on individuals wth location > 5

```{r}
Metadata_loc_above_5_ind <- Metadata_cleaned_microbiome %>%
    group_by(location.numeric) %>%
    filter(n() > 5)
```

## visualisation

### Set color Palette
```{r}
col_gir <- c("#a0be25", "#25be4b")
```

### Difference distribution of the data based on Sex

```{r}
# Create a plot to visualize the difference in nbr. individuals per population
ggplot(Metadata_cleaned_microbiome, aes(x = calc_sex, fill = calc_sex)) +
    geom_bar(stat = "count") +
    scale_fill_manual(values = col_gir) +
    theme(legend.position = "none") +
    ggtitle("Nbr. of Individuals per Sex") +
    xlab("Sex") +
    ylab("Nbr. of Individuals")
```

### Difference distribution of the data based on Island

```{r}
# Create a plot to visualize the difference in nbr. individuals per population
ggplot(Metadata_cleaned_microbiome, aes(x = Island, fill = Island)) +
    geom_bar(stat = "count") +
    scale_fill_manual(values = col_gir) +
    theme(legend.position = "none") +
    ggtitle("Nbr. of Individuals per Island") +
    xlab("Island") +
    ylab("Nbr. of Individuals")
```

### Difference in distribution of the data based on Location sampled

```{r}
# Create a plot to visualize the difference in nbr. individuals per population
ggplot(Metadata_cleaned_microbiome, aes(x = location.numeric, fill = location.numeric)) +
    geom_bar(stat = "count") +
    theme(legend.position = "none") +
    ggtitle("Nbr. of Individuals per location") +
    xlab("location category") +
    ylab("Nbr. of Individuals")
```

### Combine the distribution of Location, Island and sex
```{r}
ggplot(Metadata_loc_above_5_ind, aes(x = location.numeric, fill = location.numeric)) +
    geom_bar(stat = "count") +
    theme(legend.position = "none") +
    ggtitle("Nbr. of Individuals per location and sex") +
    xlab("location category") +
    ylab("Nbr. of Individuals") +
    facet_grid(calc_sex ~ Island, scales = "free")
```

### Difference based on Age category
```{r}
# Create a plot to visualize the difference in nbr. individuals per Age category
ggplot(Metadata_cleaned_microbiome, aes(x = Age_category, fill = Age_category)) +
    geom_bar(stat = "count") +
    scale_fill_manual(values = col_gir) +
    theme(legend.position = "none") +
    ggtitle("Nbr. of Individuals per Age Category") +
    xlab("Age Category") +
    ylab("Nbr. of Individuals")
```

###  Distribution of Body mass compared to Age category
```{r}
# Create a plot to visualize the difference in body mass per Age category
ggplot(Metadata_cleaned_microbiome, aes(x = Age_category, y = Body.mass..g., fill = Age_category)) +
    geom_violin(alpha = 1, color = "black", lwd = 0.36, width = 0.7) +
    geom_boxplot(width = 0.2, fill = "white") +
    # geom_boxplot(alpha = 1, color = "black", outlier.color = "white", outlier.size = 0, lwd= 0.36, width = 0.7) +
    geom_point(position = position_jitterdodge(jitter.width = 0.35), alpha = 1, size = 1.25, color = "black", pch = 21) +
    ylab("Body Mass (g)") +
    xlab("Age Category") +
    labs(fill = "Age Category") +
    scale_fill_manual(values = col_gir) +
    scale_color_manual(values = col_gir) +
    theme_bw() +
    theme(axis.title.x = element_text(size = 18, margin = margin(t = 7)), axis.title.y = element_text(size = 18, margin = margin(r = 8)), axis.text.y = element_text(size = 18, color = "black"), axis.text.x = element_text(size = 18, color = "black"), strip.text.x = element_blank()) +
    theme(panel.border = element_rect(colour = "grey29"), panel.grid.minor = element_blank(), panel.grid.major = element_blank())

# Statistical test kurkal-wallis
kw <- kruskal.test(Body.mass..g. ~ Age_category, data = Metadata_cleaned_microbiome)
kw
```

### Look at the correlations between environmental variable

```{r}
# Extract all possible covariates
Spearman_covariate <- Metadata_cleaned_microbiome[c(24:26, 28:29)]

# Change Age_category to binary (Juv = 1; Ad = 2)
Spearman_covariate$Age_category <- c(Adult = 2, Juvenile = 1)[Spearman_covariate$Age_category]

# Change Island to binary (Kauai = 1; Bermuda = 2)
Spearman_covariate$Island <- c(Bermuda = 2, Kauai = 1)[Spearman_covariate$Island]

# Change Sample.Type to binary (Cecum = 1; Scrape = 2)
Spearman_covariate$Sample.Type <- c(Scrape = 2, Cecum = 1)[Spearman_covariate$Sample.Type]

# Change everything to numeric
Spearman_covariate$sex_calc_numeric <- as.numeric(Spearman_covariate$sex_calc_numeric)
Spearman_covariate$location.numeric <- as.numeric(Spearman_covariate$location.numeric)

# Remove NA
Spearman_covariate <- Spearman_covariate[!is.na(Spearman_covariate$Body.mass..g.), ]

# Realise a partial Spearman correlation test
Spearman <- cor(Spearman_covariate, method = "spearman")

# Visualize the correlation matrix
corrplot(Spearman, method = "number", type = "lower", diag = FALSE)
```

## Plot Genomic PCA

```{r}
# Import the Genomic PCA 
Genome_eigenvec <- read.table("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/PLINK/PCA/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink-job9079824.PCA-job9667882.eigenvec", header = FALSE)

# Create a plot to visualize the genomic PCA
ggplot(data = Metadata_cleaned_microbiome, aes(x = PC1, y = PC2, fill = Island)) +
      theme_bw() +
      #stat_ellipse(geom = "polygon", type = "t", alpha = 0.2, level = 0.9) + # confidence interval
      geom_point(color = "black", shape = 21, size = 2.5, stroke = 0.3) 
      xlab("PC1") +
      ylab("PC2") +
      theme(
            axis.title.x = element_text(size = 13, margin = margin(t = 3)),
            axis.title.y = element_text(size = 13, margin = margin(r = 2)),
            axis.text.y = element_text(size = 11, color = "black"),
            axis.text.x = element_text(size = 11, color = "black"),
            strip.text = element_text(size = 13)
      ) +
      theme(
            panel.grid.minor = element_blank(),
            panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "grey29")
      ) +
      theme(
            legend.title = element_text(size = 13),
            legend.text = element_text(size = 13),
            legend.key.size = unit(0.6, "cm"),
            legend.position = "right"
      ) +
      #scale_color_manual(values = col_gir) +
      #scale_fill_manual(values = col_gir)

```

## PC variance explained for Genome PCA

```{r}
# Import the eigenvalue of the PCA
Genome_eigenvalue <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/PLINK/PCA/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink-job9079824.PCA-job9667882.eigenval", header = FALSE)

# visualise the PVE
plot(Genome_eigenvalue$V1,
    xlab = "Principal Component",
    ylab = "Proportion of variance explained",
    main = "variance explained by each PC",
    type = "b",
    col = "red",
    lwd = 2
)

# visualise the cumulative effect
plot(cumsum(Genome_eigenvalue$V1),
    xlab = "Principal Component",
    ylab = "Proportion of variance explained",
    main = "cumulative variance explained by PCs",
    type = "b",
    col = "blue",
    lwd = 2
)
```

## Save the  Metadata file into a workspace for safety

```{r}
# Remove all unnecessary object
rm(list = ls()[!(ls() %in% c("Metadata_cleaned_microbiome"))])

# Save the cleaned metadata object
save.image(file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Backup_workspace/Cleaned_Metadata.RData")
```