# GWAS Visualisation Script for Community metrics (Simpson/Shannon/Pielou)

This is gonna be used to visualise the output of different GWAS Analysis from Community metrics which gave significant result by doing Manhattan plot, qqplot, etc. for quality control.
## Import Packages
```{r}
library(qqman)
library(dplyr)
```


## Import the Manhattan function of Willlian
```{r, eval = F}
# Import form Willian github
source("https://github.com/williantafsilva/scripts/raw/refs/heads/main/function-plot_manhattan.R")
```

## Import the data
```{r, eval = F}
# Import the GEMMA output for different trait
GWAS_output_Simpson <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/Simpson/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.6_Community_GWAS-job9738904.assoc.txt")

GWAS_output_Shannon <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/Shannon/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.5_Community_GWAS-job9738903.assoc.txt")

GWAS_output_Pielou <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/pielou/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.9_Community_GWAS-job9729436.assoc.txt")

GWAS_output_Chao1 <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/Chao1/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.3_Community_GWAS-job9738428.assoc.txt")

GWAS_output_InvSimpson <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/InvSimpson/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.7_Community_GWAS-job9738905.assoc.txt")

GWAS_output_Observed <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/Observed/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.2_Community_GWAS-job9738427.assoc.txt")

GWAS_output_simpson_evenness <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Community/simpson_evenness/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_Community-job9729389.8_Community_GWAS-job9738952.assoc.txt")
```


## Simpson Metric GWAS

### Full

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Simpson$p_score_BH <- p.adjust(GWAS_output_Simpson$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Simpson %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Simpson %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Simpson, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Simpson$chr,
  chrpositionvector = GWAS_output_Simpson$ps,
  values = -log10(GWAS_output_Simpson$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Simpson Index"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Simpson$p_score, main = "Q-Q plot of GWAS p-values for Simpson Index", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```
#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Simpson$p_score), 1) / qchisq(0.5, 1)
```


### Without Outliers

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Simpson_wo_out$p_score_BH <- p.adjust(GWAS_output_Simpson_wo_out$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Simpson_wo_out %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Simpson_wo_out %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Simpson_wo_out, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Simpson_wo_out$chr,
  chrpositionvector = GWAS_output_Simpson_wo_out$ps,
  values = -log10(GWAS_output_Simpson_wo_out$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Simpson Index without outliers"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Simpson_wo_out$p_score, main = "Q-Q plot of GWAS p-values for Simpson Index without outliers", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```
#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Simpson_wo_out$p_score), 1) / qchisq(0.5, 1)
```

## Shannon Metric GWAS

### Full

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Shannon$p_score_BH <- p.adjust(GWAS_output_Shannon$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Shannon %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Shannon %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Shannon, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Shannon$chr,
  chrpositionvector = GWAS_output_Shannon$ps,
  values = -log10(GWAS_output_Shannon$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Shannon Index"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Shannon$p_score, main = "Q-Q plot of GWAS p-values for Shannon Index", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Shannon$p_score), 1) / qchisq(0.5, 1)
```

### Without Outliers

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Shannon_wo_out$p_score_BH <- p.adjust(GWAS_output_Shannon_wo_out$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Shannon_wo_out %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Shannon_wo_out %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Shannon_wo_out, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Shannon_wo_out$chr,
  chrpositionvector = GWAS_output_Shannon_wo_out$ps,
  values = -log10(GWAS_output_Shannon_wo_out$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Shannon Index without outliers"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Shannon_wo_out$p_score, main = "Q-Q plot of GWAS p-values for Shannon Index without outliers", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Shannon_wo_out$p_score), 1) / qchisq(0.5, 1)
```

## Pielou Metric GWAS

### Full

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Pielou$p_score_BH <- p.adjust(GWAS_output_Pielou$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Pielou %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Pielou %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Pielou, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Pielou$chr,
  chrpositionvector = GWAS_output_Pielou$ps,
  values = -log10(GWAS_output_Pielou$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Pielou Index"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Pielou$p_score, main = "Q-Q plot of GWAS p-values for Pielou Index", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Pielou$p_score), 1) / qchisq(0.5, 1)
```

### Without Outliers

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Pielou_wo_out$p_score_BH <- p.adjust(GWAS_output_Pielou_wo_out$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Pielou_wo_out %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Pielou_wo_out %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Pielou_wo_out, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Pielou_wo_out$chr,
  chrpositionvector = GWAS_output_Pielou_wo_out$ps,
  values = -log10(GWAS_output_Pielou_wo_out$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Pielou Index without outliers"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Pielou_wo_out$p_score, main = "Q-Q plot of GWAS p-values for Pielou Index without outliers", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Pielou_wo_out$p_score), 1) / qchisq(0.5, 1)
```

## Chao1 Metric GWAS

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Chao1$p_score_BH <- p.adjust(GWAS_output_Chao1$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Chao1 %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Chao1 %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Chao1, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Chao1$chr,
  chrpositionvector = GWAS_output_Chao1$ps,
  values = -log10(GWAS_output_Chao1$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Chao1 Index"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Chao1$p_score, main = "Q-Q plot of GWAS p-values for Chao1 Index", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_Chao1$p_score), 1) / qchisq(0.5, 1)
```


## InvSimpson Metric GWAS

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_InvSimpson$p_score_BH <- p.adjust(GWAS_output_InvSimpson$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_InvSimpson %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_InvSimpson %>%
  top_n(n = -1, wt = p_score)
```

#### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_InvSimpson, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_InvSimpson$chr,
  chrpositionvector = GWAS_output_InvSimpson$ps,
  values = -log10(GWAS_output_InvSimpson$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for InvSimpson Index"
)
```

#### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_InvSimpson$p_score, main = "Q-Q plot of GWAS p-values for InvSimpson Index", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

#### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1 - median(GWAS_output_InvSimpson$p_score), 1) / qchisq(0.5, 1)
```

## Observed Metric GWAS

### Full

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Observed$p_score_BH <- p.adjust(GWAS_output_Observed$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Observed %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Observed %>%
  top_n(n = -1, wt = p_score)
```

## simpson evenness Metric GWAS

### Full

#### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_simpson_evenness$p_score_BH <- p.adjust(GWAS_output_simpson_evenness$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_simpson_evenness %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_simpson_evenness %>%
  top_n(n = -1, wt = p_score)
```