# GWAS Visualisation Script for Community metrics (Simpson/Shannon/Pielou)

This is gonna be used to visualise the output of different GWAS Analysis from Community metrics which gave significant result by doing Manhattan plot, qqplot, etc. for quality control.
## Import Packages
```{r}
library(qqman)
library(dplyr)
```


## Import the Manhattan function of Willlian
```{r, eval = F}
# Import form Willian github
source("https://github.com/williantafsilva/scripts/raw/refs/heads/main/function-plot_manhattan.R")
```

## Import the data
```{r, eval = F}
# Import the GEMMA output for different trait
GWAS_output_Simpson <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Simpson/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_extra-job9425802.2_CLR_GWAS-job9425983.assoc.txt")

GWAS_output_Shannon <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Shannon/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome-job9088533._CLR_GWAS-job9280634.assoc.txt")

GWAS_output_Pielou <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Analysis/GWAS/Pielou/cc5k.autosomes.selectFERAL.renamesamples.minimize.vcftoplink.geno_microbiome_extra-job9425802.3_CLR_GWAS-job9425984.assoc.txt")
```


## Simpson Metric GWAS

### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Simpson$p_score_BH <- p.adjust(GWAS_output_Simpson$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Simpson %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Simpson %>%
  top_n(n = -1, wt = p_score)
```

### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Simpson, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Simpson$chr,
  chrpositionvector = GWAS_output_Simpson$ps,
  values = -log10(GWAS_output_Simpson$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Simpson Index"
)
```

### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Simpson$p_score, main = "Q-Q plot of GWAS p-values for Simpson Index", xlim = c(0, 7), ylim = c(0,12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```
### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1-median(GWAS_output_Simpson$p_score),1)/qchisq(0.5,1)
```

## Shannon Metric GWAS

### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Shannon$p_score_BH <- p.adjust(GWAS_output_Shannon$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Shannon %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Shannon %>%
  top_n(n = -1, wt = p_score)
```

### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Shannon, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Shannon$chr,
  chrpositionvector = GWAS_output_Shannon$ps,
  values = -log10(GWAS_output_Shannon$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Shannon Index"
)
```

### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Shannon$p_score, main = "Q-Q plot of GWAS p-values for Shannon Index", xlim = c(0, 7), ylim = c(0,12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1-median(GWAS_output_Shannon$p_score),1)/qchisq(0.5,1)
```
## Pielou Metric GWAS

### Correct the p-value for multiple testing using FDR
```{r}
# adjust the p-value using FDR
GWAS_output_Pielou$p_score_BH <- p.adjust(GWAS_output_Pielou$p_score, method = "BH")

# Check if any p-values is above the threashold of q < 0.05
nrow(GWAS_output_Pielou %>%
  filter(p_score_BH < 0.05))

# Find top SNP
GWAS_output_Pielou %>%
  top_n(n = -1, wt = p_score)
```

### Make a Manhattan plot

```{r}
## using the qqman package
manhattan(GWAS_output_Pielou, chr = "chr", bp = "ps", snp = "rs", p = "p_score", logp = TRUE)
```


```{r}
# Use Willian Manhattan plot function
plot_manhattan(
  chrnamevector = GWAS_output_Pielou$chr,
  chrpositionvector = GWAS_output_Pielou$ps,
  values = -log10(GWAS_output_Pielou$p_score_BH),
  abovethreshold = -log10(0.05),
  belowthreshold = -Inf,
  colors = c("blue", "red"),
  xlabel = "Chromosome",
  ylabel = "LogPvalue",
  plottitle = "Manhattan plot for Pielou Index"
)
```

### Q-Q Plot

```{r}
# Use the qqman package
qq(GWAS_output_Pielou$p_score, main = "Q-Q plot of GWAS p-values for Pielou Index", xlim = c(0, 7), ylim = c(0,12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```

### Calculate Genomic Inflation
```{r}
# Calculate genomic inflation factor
qchisq(1-median(GWAS_output_Pielou$p_score),1)/qchisq(0.5,1)
```