# Phenotypical trait DATA preparation

This Script is made top create the Dataframe, which is gonna be used in the GWAS.

The Dataframe will contain:
- Richness
- Eveness
- Beta-div (PC axis)
- Relative Abundance of most prevalent bacteria

```{r , eval = F}
library(readr)
```

## Create the Phenotype dataframe

```{r, eval = F}
# Add ID first
Chicken_microbiome_pheno <- as.data.frame(metadata$ID, row.names = c(sample_names(ps_rare)))
colnames(Chicken_microbiome_pheno) <- c("IID")

# Richness
rich <- estimate_richness(ps_rare)
Chicken_microbiome_pheno <- cbind(Chicken_microbiome_pheno, rich)

# Eveness
p <- microbiome::evenness(otu_table(ps_rare))
Chicken_microbiome_pheno <- cbind(Chicken_microbiome_pheno, p)

# Principal componant of beta div
ord_ps <- ordinate(ps_Norm, "PCoA", "bray")
Chicken_microbiome_pheno <- cbind(Chicken_microbiome_pheno, ord_ps$vectors[, 1:5])
colnames(Chicken_microbiome_pheno)[5:9] <- c("BRAY_PC1", "BRAY_PC2", "BRAY_PC3", "BRAY_PC4", "BRAY_PC5")

# Relative abundance of the set core microbiome
ps_ASV_GWAS <- core(ps_Norm, detection = 0, prevalence = 0.5) # 50% percent see if good threashold
ps_ASV_GWAS <- transform_sample_counts(ps_ASV_GWAS, function(x) x / sum(x))

# Transform the core microbiome in center log ratio (CLR)
ps_ASV_GWAS_CLR <- microbiome::transform(ps_ASV_GWAS, "clr")

# Add to the phenotype
Chicken_microbiome_pheno_clr <- cbind(Chicken_microbiome_pheno, otu_table(ps_ASV_GWAS_CLR))
Chicken_microbiome_pheno <- cbind(Chicken_microbiome_pheno, otu_table(ps_ASV_GWAS))


# Format the 0 value in NA ?
Chicken_microbiome_pheno[Chicken_microbiome_pheno == 0] <- NA
Chicken_microbiome_pheno_clr[Chicken_microbiome_pheno_clr == 0] <- NA

# Remove Sample ERC44: non present in genome data
Chicken_microbiome_pheno <- Chicken_microbiome_pheno[Chicken_microbiome_pheno$IID != "ERC44", ]
Chicken_microbiome_pheno_clr <- Chicken_microbiome_pheno_clr[Chicken_microbiome_pheno_clr$IID != "ERC44", ]

# Create a text file with the phenotypes
write_delim(Chicken_microbiome_pheno, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Feral_Chicken_PHENO.txt")

# Create a text file with the phenotypes (Without Headers)
write_delim(Chicken_microbiome_pheno, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Feral_Chicken_PHENO_without_headers.txt", col_names = FALSE)

write_delim(Chicken_microbiome_pheno_clr[c(1, 10:ncol(Chicken_microbiome_pheno_clr))], file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Feral_Chicken_PHENO_clr_with_headers.txt", col_names = T)
```

## New values for phenotypes
```{r}
# Add ID first
Chicken_microbiome_pheno_extra <- as.data.frame(metadata$ID, row.names = c(sample_names(ps_rare)))
colnames(Chicken_microbiome_pheno_extra) <- c("IID")

# Richness
Chicken_microbiome_pheno_extra$Simpson <- rich$Simpson

# Eveness
Chicken_microbiome_pheno_extra$Eveness <- p$pielou

# PD (Faith) Diversity
Chicken_microbiome_pheno_extra$PD <- p2$PD

# Principal componant of beta div aitchinson
ord_clr <- ordinate(ps_CLR, "PCoA", "euclidean")
Chicken_microbiome_pheno_extra <- cbind(Chicken_microbiome_pheno_extra, ord_clr$vectors[, 1:5])
colnames(Chicken_microbiome_pheno_extra)[5:9] <- c("AITCH_PC1", "AITCH_PC2", "AITCH_PC3", "AITCH_PC4", "AITCH_PC5")

# Principal componant of beta div Unifrac
ord_uni <- ordinate(ps_Norm, "PCoA", "unifrac", weighted = T)
Chicken_microbiome_pheno_extra <- cbind(Chicken_microbiome_pheno_extra, ord_uni$vectors[, 1:5])
colnames(Chicken_microbiome_pheno_extra)[10:14] <- c("UNI_PC1", "UNI_PC2", "UNI_PC3", "UNI_PC4", "UNI_PC5")

# Remove Sample ERC44: non present in genome data
Chicken_microbiome_pheno_extra <- Chicken_microbiome_pheno_extra[Chicken_microbiome_pheno_extra$IID != "ERC44", ]

# Create a text file with the phenotypes (Without Headers)
write_delim(Chicken_microbiome_pheno_extra, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Feral_Chicken_PHENO_extra_without_headers.txt", col_names = F)
```

## New values + Shannon/Simpson/Pielou without outliers
```{r}
# Add ID first
Chicken_microbiome_pheno_extra_2 <- as.data.frame(metadata$ID, row.names = c(sample_names(ps_rare)))
colnames(Chicken_microbiome_pheno_extra_2) <- c("IID")

# Add Chao1
Chicken_microbiome_pheno_extra_2$Chao1 <- rich$Chao1

# Add Inverse Simpson
Chicken_microbiome_pheno_extra_2$InvSimp <- rich$InvSimpson

# Add Simpson without outlier
Chicken_microbiome_pheno_extra_2$Simpson_wo_out <- rich$Simpson
Chicken_microbiome_pheno_extra_2$Simpson_wo_out[Chicken_microbiome_pheno_extra_2$Simpson_wo_out <= 0.85] <- NA

# Add Shannon without outliers
Chicken_microbiome_pheno_extra_2$Shannon_wo_out <- rich$Shannon
Chicken_microbiome_pheno_extra_2$Shannon_wo_out[Chicken_microbiome_pheno_extra_2$Shannon_wo_out <= 3] <- NA

# Add Pielou without outliers
Chicken_microbiome_pheno_extra_2$pielou_wo_out <- p$pielou
Chicken_microbiome_pheno_extra_2$pielou_wo_out[Chicken_microbiome_pheno_extra_2$pielou_wo_out <= 0.6] <- NA

# Remove Sample ERC44: non present in genome data
Chicken_microbiome_pheno_extra_2 <- Chicken_microbiome_pheno_extra_2[Chicken_microbiome_pheno_extra_2$IID != "ERC44", ]

# Create a text file with the phenotypes (Without Headers)
write_delim(Chicken_microbiome_pheno_extra_2, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Feral_Chicken_PHENO_extra_2_wo_outliers.txt", col_names = F)
```

## Explore Phenotype distribution
```{r}
## Phenotype Distribution visualisation
hist(Chicken_microbiome_pheno_extra$Shannon, breaks = 40)

hist(rich$Shannon, breaks = 40)
```
## Phenotype and Covariates Dataframe
```{r, eval = F}
# Load the data file
feral_covariates_microbiome <- read.delim("/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/feral_covariates_microbiome.txt")

# Reformat
feral_covariates_microbiome$Island <- replace(feral_covariates_microbiome$Island, feral_covariates_microbiome$Island == "kauai", "Kauai")
feral_covariates_microbiome$Island <- replace(feral_covariates_microbiome$Island, is.na(feral_covariates_microbiome$Island), "Kauai")

# For merging purposes change ID to IID
colnames(feral_covariates_microbiome)[1] <- c("IID")

# Merge the dataset (full)
feral_covariates_microbiome_alpha <- merge(Chicken_microbiome_pheno, feral_covariates_microbiome, by = "IID")

feral_covariates_microbiome_RUN1 <- merge(Chicken_microbiome_pheno_clr, feral_covariates_microbiome, by = "IID")

feral_covariates_microbiome_RUN1_extra <- merge(Chicken_microbiome_pheno_extra, feral_covariates_microbiome, by = "IID")

# Add Sample Type as a possible covariate
feral_covariates_microbiome_alpha$Sample.Type <- metadata$Sample.Type[metadata$ID != "ERC44"]

feral_covariates_microbiome_RUN1$Sample.Type <- metadata$Sample.Type[metadata$ID != "ERC44"]

feral_covariates_microbiome_RUN1_extra$Sample.Type <- metadata$Sample.Type[metadata$ID != "ERC44"]

# Remove non important covariants (calc_sex)
feral_covariates_microbiome_RUN1 <- feral_covariates_microbiome_RUN1[, !(names(feral_covariates_microbiome_RUN1) %in% c("calc_sex", "location.numeric", "Latitude", "Longitude", "latitude.longitude"))]

# Change Age_category to binary (Juv = 1; Ad = 2)
feral_covariates_microbiome_RUN1$Age_category <- c(Adult = 2, Juvenile = 1)[feral_covariates_microbiome_RUN1$Age_category]

# Change Island to binary (Kauai = 1; Bermuda = 2)
feral_covariates_microbiome_RUN1$Island <- c(Bermuda = 2, Kauai = 1)[feral_covariates_microbiome_RUN1$Island]

# Change Sample.Type to binary (Cecum = 1; Scrape = 2)
feral_covariates_microbiome_RUN1$Sample.Type <- c(Scrape = 2, Cecum = 1)[feral_covariates_microbiome_RUN1$Sample.Type]
```

## Linear Model to pick Covariate for each Phenotype (look if gonna be used or not)
```{r, eval = F}
feral_covariates_microbiome_alpha$norm_Simpson <- scale(feral_covariates_microbiome_alpha$Simpson)

# Linear model (Sample Type)
oneway.model <- lm(pielou ~ Island, data = feral_covariates_microbiome_alpha)
summary(oneway.model)

feral_covariates_microbiome_alpha$residual <- oneway.model$residual

qplot(oneway.model$residual,
    geom = "histogram",
    bins = 10
) +
    labs(
        title = "Histogram of residuals",
        x = "residual"
    )

ggplot(
    data = feral_covariates_microbiome_alpha,
    aes(sample = residual)
) +
    geom_qq() +
    geom_qq_line(colour = "red") +
    labs(title = "Quantile plot of residuals")
# Linear model (PC 14)
oneway.model <- lm(Richness ~ PC14, data = feral_covariates_microbiome_RUN1_Richness)
summary(oneway.model)

# Linear model (PC 15)
oneway.model <- lm(Richness ~ PC15, data = feral_covariates_microbiome_RUN1_Richness)
summary(oneway.model)

# Linear model (PC 19)
oneway.model <- lm(Richness ~ PC19, data = feral_covariates_microbiome_RUN1_Richness)
summary(oneway.model)
```

## Create the covariate dataframes

```{r, eval = F}
# General covariates file
write_delim(feral_covariates_microbiome_RUN1[c(1, (ncol(Chicken_microbiome_pheno) + 1):ncol(feral_covariates_microbiome_RUN1))], file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Chicken_microbiome_cov_without_location.txt", col_names = F)

# Covariate file without the PCs
feral_covariates_microbiome_Without_PC <- feral_covariates_microbiome_RUN1[, !(names(feral_covariates_microbiome_RUN1) %in% c("calc_sex", "location.numeric", "Latitude", "Longitude", "latitude.longitude", "PC1", "PC2", "PC3", "PC4", "PC5", "PC6", "PC7", "PC8", "PC9", "PC10", "PC11", "PC12", "PC13", "PC14", "PC15", "PC16", "PC17", "PC18", "PC19", "PC20"))]
write_delim(feral_covariates_microbiome_Without_PC[c(1, (ncol(Chicken_microbiome_pheno) + 1):ncol(feral_covariates_microbiome_Without_PC))], file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Chicken_microbiome_cov_without_PC_and_Loc.txt", col_names = F)

# Covariate file with the first 10 PCs
feral_covariates_microbiome_With_10PC <- feral_covariates_microbiome_RUN1[, !(names(feral_covariates_microbiome_RUN1) %in% c("calc_sex", "location.numeric", "Latitude", "Longitude", "latitude.longitude", "PC11", "PC12", "PC13", "PC14", "PC15", "PC16", "PC17", "PC18", "PC19", "PC20"))]
write_delim(feral_covariates_microbiome_With_10PC[c(1, (ncol(Chicken_microbiome_pheno) + 1):ncol(feral_covariates_microbiome_With_10PC))], file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Chicken_microbiome_cov_with_10PC_and_without_Loc.txt", col_names = F)

# Covariates file for Richness
# keep important covariants
feral_covariates_microbiome_Richness <- feral_covariates_microbiome_RUN1[, (names(feral_covariates_microbiome_RUN1) %in% c("IID", "Sample.Type", "PC14", "PC15", "PC19", "Island"))]

write_delim(feral_covariates_microbiome_Richness, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/Chicken_microbiome_cov_Richness.txt", col_names = F)
```

## Create an ID list for Genome dataset filtering
```{r, eval = F}
ID_feral <- as.data.frame(feral_covariates_microbiome_RUN1$IID)

write_delim(ID_feral, file = "/Users/matiasbeckerburgos/Desktop/Master_in_Bioinformatics/Master_Thesis/Hawaiian_feral_Chicken_Microbiome/Data/GWAS_DATA/ID_Feral_Chicken.txt", col_names = F)
```

## Save the new created object

```{r, eval = F}
# Erase all objects except metadata, Chicken_Microbiome_pheno, ps_object, save
rm(list = ls()[!(ls() %in% c("metadata", "ps", "save", "Chicken_microbiome_pheno", "ps_Norm", "ps_Norm_F", "ps_core", "ps_rare"))])

save.image(file.path(save, "GWAS_Data_RUN1.RData"))
```